File: netra-app/src/lib/quality_check.py
"""
Netra Step 2: Fundus Image Quality Check (Python OpenCV Reference Implementation)
Runs 4 clinical quality checks before DR AI inference:
1. FOV Masking & Coverage Ratio (CR_FOV)
2. Focus Score via Modified Laplacian Variance (Var(Laplacian))
3. Illumination Uniformity (Mean Green Channel Luminance)
4. Saturation & Glare Ratio
"""

import cv2
import numpy as np
from typing import Dict, Any, List

def analyze_fundus_quality(image_path_or_bytes) -> Dict[str, Any]:
    # 1. Load Image
    if isinstance(image_path_or_bytes, str):
        img = cv2.imread(image_path_or_bytes)
    else:
        nparr = np.frombuffer(image_path_or_bytes, np.uint8)
        img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)

    if img is None:
        return {
            "gradable": False,
            "quality_status": "rejected",
            "rejection_reasons": ["Invalid or corrupt image file"],
        }

    # Resize to standard width 512 for fast edge execution
    h, w = img.shape[:2]
    target_w = 512
    scale = target_w / float(w)
    img_resized = cv2.resize(img, (target_w, int(h * scale)), interpolation=cv2.INTER_AREA)
    
    # Extract Green Channel (Highest contrast for retinal structures)
    green_ch = img_resized[:, :, 1]
    
    # 2. Compute FOV Mask (Threshold background pixels)
    gray = cv2.cvtColor(img_resized, cv2.COLOR_BGR2GRAY)
    _, fov_mask = cv2.threshold(gray, 20, 255, cv2.THRESH_BINARY)
    
    fov_pixels = cv2.countNonZero(fov_mask)
    total_pixels = img_resized.shape[0] * img_resized.shape[1]
    fov_coverage_ratio = fov_pixels / float(total_pixels)

    if fov_pixels == 0:
        return {
            "gradable": False,
            "quality_status": "rejected",
            "rejection_reasons": ["No retinal field detected (Black Image)"],
        }

    # 3. Illumination & Glare Check
    mean_illumination = cv2.mean(green_ch, mask=fov_mask)[0]
    
    # Saturated glare pixels (intensity > 245 inside FOV)
    _, glare_mask = cv2.threshold(green_ch, 245, 255, cv2.THRESH_BINARY)
    glare_mask = cv2.bitwise_and(glare_mask, fov_mask)
    glare_pixels = cv2.countNonZero(glare_mask)
    glare_ratio = glare_pixels / float(fov_pixels)

    # 4. Focus Score (Laplacian Variance inside FOV)
    laplacian = cv2.Laplacian(green_ch, cv2.CV_64F)
    # Mask out background to prevent edge boundary artificially inflating score
    kernel = np.ones((5, 5), np.uint8)
    eroded_fov = cv2.erode(fov_mask, kernel, iterations=2)
    
    laplacian_fov = laplacian[eroded_fov > 0]
    if len(laplacian_fov) > 0:
        focus_score = float(np.var(laplacian_fov))
    else:
        focus_score = 0.0

    # 5. Clinical Rejection Thresholds
    rejection_reasons: List[str] = []

    if fov_coverage_ratio < 0.35:
        rejection_reasons.append("Insufficient Retinal Field of View (<35% of frame)")
    
    if focus_score < 60.0:
        rejection_reasons.append(f"Out of Focus / Blurry (Laplacian Variance: {focus_score:.1f} < 60.0)")

    if mean_illumination < 25.0:
        rejection_reasons.append(f"Severe Under-Exposure (Mean Luminance: {mean_illumination:.1f} < 25.0)")
    elif mean_illumination > 225.0:
        rejection_reasons.append(f"Severe Over-Exposure / Flash Saturation ({mean_illumination:.1f} > 225.0)")

    if glare_ratio > 0.08:
        rejection_reasons.append(f"Corneal Glare / Flash Reflection (>8% saturated pixels)")

    gradable = len(rejection_reasons) == 0
    if gradable:
        quality_status = "passed" if focus_score >= 100.0 and glare_ratio <= 0.03 else "warning"
    else:
        quality_status = "rejected"

    return {
        "gradable": gradable,
        "quality_status": quality_status,
        "rejection_reasons": rejection_reasons,
        "fov_coverage_ratio": round(fov_coverage_ratio, 3),
        "focus_score": round(focus_score, 1),
        "mean_illumination": round(mean_illumination, 1),
        "glare_ratio": round(glare_ratio, 3),
        "retina_visible": fov_coverage_ratio >= 0.35 and mean_illumination >= 25.0,
    }

if __name__ == "__main__":
    print("Step 2 Quality Check Engine initialized.")

File: netra-app/src/lib/qualityCheck.ts
/**
 * Netra Step 2: Fundus Image Quality Check Engine
 * Calculates real-time clinical quality metrics on fundus photos via HTML5 Canvas.
 */

export interface QualityMetrics {
  gradable: boolean;
  qualityStatus: 'passed' | 'warning' | 'rejected';
  rejectionReasons: string[];
  fovCoverageRatio: number; // 0.0 - 1.0 (Target: >= 0.45)
  focusScore: number;       // Laplacian Variance (Target: >= 120.0)
  meanIllumination: number; // 0 - 255 (Target: 35.0 - 210.0)
  glareRatio: number;       // 0.0 - 1.0 (Target: <= 0.04)
  retinaVisible: boolean;
}

/**
 * Analyzes an HTMLImageElement using canvas pixel data to run Step 2 Quality Gate
 */
export function analyzeFundusQuality(imgElement: HTMLImageElement): QualityMetrics {
  // Fallback for when image is unreadable / CORS-blocked / zero-sized.
  // Return gradable=false so the UI shows a warning instead of fake passing metrics.
  const fallbackMetrics: QualityMetrics = {
    gradable: false,
    qualityStatus: 'warning',
    rejectionReasons: ['Image quality analysis unavailable — CORS or decode error'],
    fovCoverageRatio: 0,
    focusScore: 0,
    meanIllumination: 0,
    glareRatio: 0,
    retinaVisible: false,
  };

  if (!imgElement || imgElement.naturalWidth === 0 || imgElement.naturalHeight === 0) {
    return fallbackMetrics;
  }

  try {
    const canvas = document.createElement('canvas');
    const ctx = canvas.getContext('2d', { willReadFrequently: true });

    const imgWidth = imgElement.naturalWidth || imgElement.width || 512;
    const imgHeight = imgElement.naturalHeight || imgElement.height || 512;

    // Downsample to max 512px for rapid real-time computation
    const scale = Math.min(1.0, 512 / Math.max(imgWidth, imgHeight));
    const width = Math.max(1, Math.floor(imgWidth * scale));
    const height = Math.max(1, Math.floor(imgHeight * scale));

    canvas.width = width;
    canvas.height = height;

    if (!ctx) return fallbackMetrics;

    ctx.drawImage(imgElement, 0, 0, width, height);
    const imageData = ctx.getImageData(0, 0, width, height);
    const data = imageData.data;

  // 1. Extract Green Channel & Compute FOV Mask
  let retinaPixelCount = 0;
  let totalLuminanceSum = 0;
  let saturatedPixelCount = 0;

  const greenChannel = new Float32Array(width * height);

  for (let i = 0; i < data.length; i += 4) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];
    const idx = i / 4;

    greenChannel[idx] = g;

    // Retinal FOV pixel condition: non-black background (R+G+B > 25)
    const isRetinaPixel = (r + g + b) > 25;
    if (isRetinaPixel) {
      retinaPixelCount++;
      totalLuminanceSum += g;
      if (g > 242 || (r > 245 && g > 245 && b > 245)) {
        saturatedPixelCount++;
      }
    }
  }

  const totalPixels = width * height;
  const fovCoverageRatio = retinaPixelCount / totalPixels;
  const meanIllumination = retinaPixelCount > 0 ? totalLuminanceSum / retinaPixelCount : 0;
  const glareRatio = retinaPixelCount > 0 ? saturatedPixelCount / retinaPixelCount : 0;

  // 2. Compute Focus Score via 3x3 Discrete Laplacian Variance on Green Channel
  // Kernel: [0, 1, 0; 1, -4, 1; 0, 1, 0]
  let laplacianSum = 0;
  let laplacianSqSum = 0;
  let validLaplacianCount = 0;

  for (let y = 1; y < height - 1; y++) {
    for (let x = 1; x < width - 1; x++) {
      const idx = y * width + x;

      // Skip background pixels outside FOV
      if (greenChannel[idx] < 15) continue;

      const top = greenChannel[(y - 1) * width + x];
      const bottom = greenChannel[(y + 1) * width + x];
      const left = greenChannel[y * width + (x - 1)];
      const right = greenChannel[y * width + (x + 1)];
      const center = greenChannel[idx];

      const lapVal = top + bottom + left + right - (4 * center);

      laplacianSum += lapVal;
      laplacianSqSum += lapVal * lapVal;
      validLaplacianCount++;
    }
  }

  let focusScore = 150.0;
  if (validLaplacianCount > 0) {
    const meanLap = laplacianSum / validLaplacianCount;
    const varianceLap = (laplacianSqSum / validLaplacianCount) - (meanLap * meanLap);
    focusScore = Math.max(0, varianceLap * 10.0); // Scaled Laplacian variance
  }

  // 3. Clinical Rejection Rules
  const rejectionReasons: string[] = [];

  if (fovCoverageRatio < 0.35) {
    rejectionReasons.push('Insufficient Retinal Field of View (<35% of frame)');
  }
  if (focusScore < 25.0) {
    rejectionReasons.push('Out of Focus / Motion Blur detected (Laplacian Variance < 25)');
  }
  if (meanIllumination < 25.0) {
    rejectionReasons.push('Severe Under-Exposure (Dark unreadable fundus)');
  } else if (meanIllumination > 225.0) {
    rejectionReasons.push('Severe Over-Exposure (Flash saturation)');
  }
  if (glareRatio > 0.08) {
    rejectionReasons.push('Corneal Glare / Flash Reflection Artifact (>8% saturated pixels)');
  }

  const gradable = rejectionReasons.length === 0;
  const qualityStatus: 'passed' | 'warning' | 'rejected' = gradable
    ? (focusScore < 100 || glareRatio > 0.03 ? 'warning' : 'passed')
    : 'rejected';

    return {
      gradable,
      qualityStatus,
      rejectionReasons,
      fovCoverageRatio: Number(fovCoverageRatio.toFixed(3)),
      focusScore: Number(focusScore.toFixed(1)),
      meanIllumination: Number(meanIllumination.toFixed(1)),
      glareRatio: Number(glareRatio.toFixed(3)),
      retinaVisible: fovCoverageRatio >= 0.35 && meanIllumination >= 25.0,
    };
  } catch (err) {
    console.warn('Canvas Quality Analysis caught error (falling back):', err);
    return fallbackMetrics;
  }
}
